### PR TITLE
dsa(indexer_backward): restore `range_constexpr` in 8 `kernel_gemm` epilogue loops

### LINKED ISSUE #548: [Perf] SM100 DSA indexer backward: `kernel_gemm` pure-kernel latency regressed 17.5-25.4% after `range_constexpr` -> `range(..., unroll_full=True)` migration (B200, DSL 4.6.1)
## Summary

The SM100 DSA indexer backward main kernel (`kernel_gemm` in
`python/cudnn/deepseek_sparse_attention/indexer_backward/indexer_backward_sm100.py`)
runs with 17.5-25.4% higher pure-kernel latency (shape-dependent) on current
`develop` than the same tree with 8 loops restored to `cutlass.range_constexpr(...)`,
on B200 (sm_100a) with `nvidia-cutlass-dsl` 4.6.1. At the anchor shape the restored
build matches the released `nvidia-cudnn-frontend` 1.26.0 wheel build of the same
kernel within 0.2%.

The observed gap is isolated to a loop-unroll API migration: 8 loops in the
`kernel_gemm` compute/reduce epilogues that the 1.26.0 wheel builds with
`cutlass.range_constexpr(...)` are built on `develop` with
`cutlass.range(..., unroll_full=True)`. Both forms fully unroll the loop body:
`range_constexpr` unrolls at trace time (static Python loop), while
`range(..., unroll_full=True)` lowers through the DSL structured-control-flow
path with a full-unroll attribute. In this configuration the migrated form
produced the higher latencies below; reverting exactly these 8 loops (and
nothing else) recovers the observed gap within measurement variation. (We did
not compare generated IR/SASS; the numbers below are the evidence.)

## Numbers

Canonical seeded workload: B=1, H=64, D=128, S_k=4096, bf16 q/k/w, int32 topk
ids, fp32 softmax grad signal fed directly to `kernel_gemm` (kernel 1,
`kernel_score_grad`, is bypassed by the harness and is not part of this
timing). Metric: nsys pure-kernel median over N=60 iterations per variant,
per-iteration pre-zeroed fp32 `d_index_k` buffers staged outside the capture
window, 2-call warmup. Primary measurement: both variants interleaved
per-iteration in ONE nsys session (anchors separated by kernel symbol;
launch-order parity cross-check agrees).

Anchor shape S_q=8192, topk=1024:

| build | kernel_gemm med (N=60) | develop / build |
|---|---|---|
| `develop` @ 11c16ff (`range(..., unroll_full=True)`) | 2056.9 us | 1.000 |
| `develop` + 8-loop revert to `range_constexpr` | 1688.8 us | 1.218 |

Both ratio directions spelled out: `develop` has **21.8% higher latency** than
the reverted build (2056.9 / 1688.8 = 1.218); the reverted build has **17.9%
lower latency** than `develop`.

Full shape matrix, same interleaved methodology (median over N=60 per variant,
std in parentheses, us):

| shape (S_k=4096) | `develop` med (std) | reverted med (std) | develop / reverted |
|---|---|---|---|
| S_q=4096 topk=512  | 576.1 (1.8)  | 490.3 (1.1)  | 1.175 |
| S_q=4096 topk=1024 | 1037.1 (2.7) | 854.1 (1.0)  | 1.214 |
| S_q=4096 topk=1536 | 1501.2 (2.9) | 1218.8 (1.5) | 1.232 |
| S_q=4096 topk=2048 | 1968.8 (3.0) | 1585.6 (1.9) | 1.242 |
| S_q=8192 topk=512  | 1137.6 (2.3) | 964.7 (1.1)  | 1.179 |
| S_q=8192 topk=1024 | 2056.9 (2.6) | 1688.8 (1.3) | 1.218 |
| S_q=8192 topk=1536 | 2981.0 (2.8) | 2405.5 (2.1) | 1.239 |
| S_q=8192 topk=2048 | 3916.1 (2.8) | 3122.4 (2.1) | 1.254 |

Reference anchors at S_q=8192/topk=1024 (separate isolated sessions, same
harness): `nvidia-cudnn-frontend` 1.26.0 wheel 1691.8 us; public `develop`
commit 3fcee36 2052.5 us; the same 3fcee36 tree with only these loops reverted
(source-only diff) 1690.9 us.

## Isolation method

- The default-kernel time is insensitive to the two grad-signal recipes tested
  (<0.2% between a raw softmax g and a kernel-1-transformed g); those two
  recipes do not explain the gap.
- The 1.26.0 wheel and current `develop` differ in this file in only two ways:
  the one-shot barrier fix of #426, and the `range_constexpr` ->
  `range(..., unroll_full=True)` migration. A source tree at the public
  `develop` commit 3fcee36 containing #426 but with the loops reverted to
  `range_constexpr` (source-only diff) measures 1690.9 us - i.e. #426 costs
  nothing measurable and the migration accounts for the observed gap within
  measurement variation.
- The migration was introduced by import commit `3a9ed3f`, which brought the
  kernel into the public tree already using `range(..., unroll_full=True)`; the
  released 1.26.0 wheel predates that commit and builds these loops with
  `range_constexpr`.

The 8 sites on `develop` @ 11c16ff
(`python/cudnn/deepseek_sparse_attention/indexer_backward/indexer_backward_sm100.py`):

| line | loop |
|---|---|
| 1077 | `dw_accum` zero-init (compute-warpgroup epilogue) |
| 1132 | `tSrS` fp32 -> bf16 convert (dS epilogue) |
| 1136 | dS scatter to `sdS_gemm_view_0` (even block) |
| 1141 | dS scatter to `sdS_gemm_view_1` (odd block) |
| 1162 | dQ TMEM readback convert x sm_scale -> bf16 |
| 1172 | dQ scatter to `sdQ_gemm_view` |
| 1191 | per-head partial reduction over `dw_accum` (dW epilogue) |
| 1260 | dK pair drain, TMEM readback -> fp32 atomicAdd (reduce warpgroup) |

## The analogous ninth site (measured; not part of the gap)

`kernel_score_grad` (kernel 1) contains one more loop with the same migration
pattern (line 1530 on `develop` @ 11c16ff, a block reduction over 128
per-thread partials). Measured separately with the same interleaved
methodology (`kernel_score_grad` pure-kernel median, N=100 per variant),
reverting it makes no difference: 49.70 us (`develop`) vs 49.73 us (reverted)
at S_q=8192/topk=1024, and 16.00 us vs 15.94 us at S_q=4096/topk=512 - both
deltas within one standard deviation (0.4-0.9 us). That site is therefore
excluded from the proposed fix; the 8 `kernel_gemm` loops above account for
the entire measured result.

## Compile-time tradeoff

`range_constexpr` trades JIT compile time for this runtime win: the DSL 4.6.1
documentation warns that fully trace-time-unrolled loops can be slow to
compile. Measured time-to-first-call with a cold DSL JIT cache (fresh process,
fresh cache dir, 3 runs per variant, anchor shape): `develop` 3.49-3.52 s vs
reverted 5.20-5.26 s, i.e. about +1.7 s one-time JIT cost per compiled shape.
At 368 us saved per call this amortizes after roughly 4600 calls.

## Reproduction

```python
# repro.py - canonical construction, default backend kernel_gemm only
import sys, torch

SQ, SK, TOPK, N = 8192, 4096, 1024, 60
B, H, D = 1, 64, 128
dev = "cuda"
torch.manual_seed(0)
q = (torch.randn(B, SQ, H, D, device=dev) * 0.1).to(torch.bfloat16).contiguous()
k = (torch.randn(B, SK, D, device=dev) * 0.1).to(torch.bfloat16).contiguous()
w = (torch.randn(B, SQ, H, device=dev) * 0.5).to(torch.bfloat16).contiguous()
idx = torch.rand(B, SQ, SK, device=dev).argsort(-1)[..., :TOPK].to(torch.int32).contiguous()
g = torch.softmax(torch.randn(B, SQ, TOPK, device=dev), -1).float().contiguous()

from cudnn.deepseek_sparse_attention.indexer_backward.indexer_backward_sm100 import indexer_backward_sm100
rc = indexer_backward_sm100(B, SQ, SK, H, D, TOPK, sm_scale=1.0, block_I=128, topk_indices_global=False)
dq = torch.empty_like(q); dw = torch.empty_like(w)

for _ in range(2):
    rc.gemm_only(q, w, k, dq, dw, torch.zeros(B, SK, D, dtype=torch.float32, device=dev), g, idx)
torch.cuda.synchronize()
dks = [torch.zeros(B, SK, D, dtype=torch.float32, device=dev) for _ in range(N)]
torch.cuda.synchronize()

torch.cuda.cudart().cudaProfilerStart()
for i in range(N):
    rc.gemm_only(q, w, k, dq, dw, dks[i], g, idx)
torch.cuda.synchronize()
torch.cuda.cudart().cudaProfilerStop()
```

```bash
# run against the repo tree (PYTHONPATH) or against the 1.26.0 wheel for the anchor
nsys profile --capture-range=cudaProfilerApi --capture-range-end=stop -t cuda,nvtx \
  -o out --force-overwrite true python repro.py
nsys stats --report cuda_gpu_kern_sum --format csv -o out_stats out.nsys-rep
# read the Med (ns) column of the kernel_cutlass_kernel_gemm_... row
```

Note: this script reproduces one build per session (single-build anchor
reproducer). The headline tables above come from a two-tree variant of the
same harness with both builds loaded into one process and interleaved
per-iteration inside one capture window; isolated single-build sessions
reproduce the same medians within ~0.3%.

## Environment

- GPU: NVIDIA B200 (sm_100a), driver 590.48.01
- CUDA 13.3, nsys 2026.3.1
- `nvidia-cutlass-dsl` 4.6.1
- torch 2.13.0a0 (NGC 26.06), python 3.12
- cudnn-frontend `develop` @ 11c16ff

## Notes

- No numerical regression was observed from the revert: `d_index_q` /
  `d_weights` are bitwise equal across builds at both tested shapes
  (S_q=8192/topk=1024 and S_q=4096/topk=512). `d_index_k` is accumulated with
  fp32 atomicAdd and is nondeterministic run-to-run; the observed cross-build
  difference (max_abs 6.7e-08, rms_rel 3.9e-07 at the anchor shape) is
  comparable in scale to same-build rerun differences (develop-develop
  max_abs 6.7e-08 / rms_rel 3.1e-07; reverted-reverted 6.0e-08 / 3.3e-07).
- If `range(..., unroll_full=True)` is intended to fully replace
  `range_constexpr`, this case may be useful as a codegen-quality reproducer
  for the DSL team as well; restoring `range_constexpr` in these 8 loops
  recovers the performance today without affecting #426.

### PR DESCRIPTION
**Pull request by zkyue**
Fixes #548 (SM100 DSA indexer backward: `kernel_gemm`
pure-kernel latency regressed 17.5-25.4% after the `range_constexpr` ->
`range(..., unroll_full=True)` migration).

## What

Reverts exactly 8 loops in the `kernel_gemm` compute/reduce epilogues of
`python/cudnn/deepseek_sparse_attention/indexer_backward/indexer_backward_sm100.py`
from `cutlass.range(..., unroll_full=True)` back to
`cutlass.range_constexpr(...)`. No other changes; the one-shot barrier fix
from #426 is untouched. An analogous ninth migration site in
`kernel_score_grad` was measured separately and shows no runtime difference
(49.70 us vs 49.73 us at the anchor shape), so it is intentionally left
unchanged - see the issue.

Both forms fully unroll the loop body: `range_constexpr` unrolls at trace
time, while `range(..., unroll_full=True)` lowers through the DSL
structured-control-flow path. In this configuration the migrated form
produced the higher latencies in the issue; this PR restores the
pre-migration form for exactly these 8 loops. See the issue for the full
isolation trail (grad-signal-recipe insensitivity, 1.26.0 wheel anchor, and a
single-migration revert probe on the public commit 3fcee36).

## Verification (B200 sm_100a, driver 590.48.01, CUDA 13.3, cutlass-dsl 4.6.1)

Performance - canonical seeded harness (B=1, H=64, D=128, S_k=4096), nsys
pure-kernel median, N=60 per variant, both variants interleaved per-iteration
in one profiling session; anchor shape S_q=8192/topk=1024:

| build | kernel_gemm med | develop / build |
|---|---|---|
| develop @ 11c16ff | 2056.9 us | 1.000 |
| this PR | 1688.8 us | **1.218** |
| 1.26.0 wheel (reference anchor, isolated session) | 1691.8 us | - |

Stated in both directions: develop has **21.8% higher latency** than this PR;
this PR has **17.9% lower latency** than develop. Across S_q {4096, 8192} x
topk {512, 1024, 1536, 2048} the same harness measures develop 17.5-25.4%
higher at every shape (full table with standard deviations in the issue).

Compile-time tradeoff (disclosed): `range_constexpr` raises time-to-first-call
with a cold DSL JIT cache from ~3.5 s to ~5.2 s at the anchor shape (one-time
per compiled shape); at 368 us saved per call it amortizes after ~4600 calls.

Numerics - no numerical regression observed:

- `d_index_q`, `d_weights` (kernel_gemm): **bitwise identical** to develop on
  the canonical harness at both S_q=8192/topk=1024 and S_q=4096/topk=512.
- `d_index_k` uses fp32 atomicAdd and is nondeterministic run-to-run: the
  develop-vs-PR difference (max_abs 6.7e-08, rms_rel 3.9e-07 at the anchor
  shape) is comparable in scale to same-build rerun differences
  (develop-develop max_abs 6.7e-08 / rms_rel 3.1e-07; PR-PR 6.0e-08 /
  3.3e-07).

Tests / lint:

- `test/python/fe_api/dsa/test_DSA_indexer_backward.py` and
  `test_DSA_dense_indexer_backward.py`: 3 passed.
- Full `fe_api/dsa/`: 81 passed, 7 skipped, 2 pre-existing failures in
  `test_DSA_indexer_top_k.py` that reproduce identically on unmodified develop
  (unrelated module).
- `pre-commit run` (pinned config: black 26.3.1, line-length 160): passed.

## Summary by CodeRabbit

* **Performance**
  * Improved compile-time loop handling for sparse attention indexing operations.
  * Preserved existing computation, memory access, synchronization, and reduction behavior.

### CODE REVIEW

**Comment by Anerudhan** [comment]
@cudnn-ci-bot run oss

**Review by Anerudhan — APPROVED** [review]
